File: interviewbit-heaps-and-maps-magician-and-chocolates.py
import heapq

class Solution:
    # @param A : integer
    # @param B : list of integers
    # @return an integer
    def nchoc(self, A, B):
        """
        Python's heapq data structure is a minheap, so I had to put the values in as value * -1,
        and then just convert back and forth when popping/pushing from/to the heap.
        """
        heap = [-num for num in B]
        heapq.heapify(heap)
        total = 0
        for time in range(A):
            if not heap:
                # if all the chocolates are already gone, or were never there, no more time needed
                break
            candies = -heapq.heappop(heap)
            total += candies
            heapq.heappush(heap, -(candies / 2))

        return total % 1000000007

File: interviewbit-heaps-and-maps-magician-and-chocolates.java
public class Solution {
    public int nchoc(int A, ArrayList<Integer> B) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((x, y) -> y - x);
        for (int i = 0; i < B.size(); i++) {
            maxHeap.add(B.get(i));
        }
        long howMany = 0;
        int bag;
        for (int i = 0; i < A; i++) {
            bag = maxHeap.poll();
            howMany += bag;
            maxHeap.add(bag / 2);
        }
        return (int) (howMany % 1000000007);
    }
}
